src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -0,0 +1,1146 @@
+        }
+
+        public ProjectId Id { get; }
+
+        private void ChangeProjectProperty<T>(ref T field, T newValue, Func<Solution, Solution> withNewValue, Action<Workspace> changeValue)
+        {
+            lock (_gate)
+            {
+                // If nothing is changing, we can skip entirely
+                if (object.Equals(field, newValue))
+                {
+                    return;
+                }
+
+                field = newValue;
+
+                if (_activeBatchScopes > 0)
+                {
+                    _projectPropertyModificationsInBatch.Add(withNewValue);
+                }
+                else
+                {
+                    _workspace.ApplyChangeToWorkspace(changeValue);
+                }
+            }
+        }
+
+        private void ChangeProjectOutputPath(ref string field, string newValue, Func<Solution, Solution> withNewValue, Action<Workspace> changeValue)
+        {
+            lock (_gate)
+            {
+                // Skip if nothing changing
+                if (field == newValue)
+                {
+                    return;
+                }
+
+                if (field != null)
+                {
+                    _workspace.RemoveProjectOutputPath(Id, field);
+                }
+
+                if (newValue != null)
+                {
+                    _workspace.AddProjectOutputPath(Id, newValue);
+                }
+
+                ChangeProjectProperty(ref field, newValue, withNewValue, changeValue);
+            }
+        }
+
+        public string AssemblyName
+        {
+            get => _assemblyName;
+            set => ChangeProjectProperty(
+                      ref _assemblyName,
+                      value,
+                       s => s.WithProjectAssemblyName(Id, value),
+                       w => w.OnAssemblyNameChanged(Id, value));
+        }
+
+        public CompilationOptions CompilationOptions
+        {
+            get => _compilationOptions;
+            set => ChangeProjectProperty(
+                       ref _compilationOptions,
+                       value,
+                       s => s.WithProjectCompilationOptions(Id, value),
+                       w => w.OnCompilationOptionsChanged(Id, value));
+        }
+
+        public ParseOptions ParseOptions
+        {
+            get => _parseOptions;
+            set => ChangeProjectProperty(
+                       ref _parseOptions,
+                       value,
+                       s => s.WithProjectParseOptions(Id, value),
+                       w => w.OnParseOptionsChanged(Id, value));
+        }
+
+        /// <summary>
+        /// The path to the output in obj.
+        /// </summary>
+        /// <remarks>This is internal for now, as it's only consumed by <see cref="EditAndContinue.VsENCRebuildableProjectImpl"/>
+        /// which directly takes a <see cref="VisualStudioProject"/>.</remarks>
+        internal string IntermediateOutputFilePath
```

Comment:
>internal [](start = 8, length = 8)

It might be currently only used by EnC but this seems generally useful property.

Reply:
Other than for ENC, it's an actively evil property, because people might use it. :smile: Lots of people try using this when they really want one of the other output paths and so they get confused. It also risky to use because many build processes are extended to where additional rewriting/modification happens on this binary before being copied to bin. I recognize that's why you *do* want it for ENC, but ironically the thing that makes it suitable for ENC's purpose makes it not suitable for most (all?) other purposes.

Reply:
In any case, the 'internal' here is in some sense moot: nobody can get a VisualStudioProject unless they created it (or the person who made it handed it out), so really only internal-to-Roslyn can use it. I anticipate that once this PR is in we'll do an API pass of VisualStudioProject and come to a conclusion here and either make this public (and also propogate the information through the workspace for public consumers), or keep this internal for ENC's sake.